libvast/test/arrow_extension_types.cpp

```diff
@@ -66,3 +70,103 @@ TEST(subnet type serde roundtrip) {
+  CHECK(caf::get_if<std::shared_ptr<arrow::NullType>>(&n));
+  CHECK(!caf::get_if<std::shared_ptr<arrow::Int64Type>>(&n));
+  CHECK(caf::get_if<std::shared_ptr<vast::enum_extension_type>>(&et));
+  CHECK(!caf::get_if<std::shared_ptr<vast::enum_extension_type>>(&pt));
+  CHECK(!caf::get_if<std::shared_ptr<vast::pattern_extension_type>>(&et));
+  CHECK(caf::get_if<std::shared_ptr<vast::pattern_extension_type>>(&pt));
+}
+
+template <class Builder, class T = typename Builder::value_type>
+std::shared_ptr<arrow::Array> makeArrowArray(std::vector<T> xs) {
+  Builder b{};
+  CHECK(b.AppendValues(xs).ok());
+  return b.Finish().ValueOrDie();
+}
+
+std::shared_ptr<arrow::Array> makeAddressArray() {
+  arrow::FixedSizeBinaryBuilder b{arrow::fixed_size_binary(16)};
+  return std::make_shared<vast::address_array>(vast::make_arrow_address(),
+                                               b.Finish().ValueOrDie());
+}
+
+TEST(arrow::Array sum type) {
+  auto str_arr = makeArrowArray<arrow::StringBuilder, std::string>({"a", "b"});
+  auto uint_arr = makeArrowArray<arrow::UInt64Builder>({7, 8});
+  auto int_arr = makeArrowArray<arrow::Int64Builder>({3, 2, 1});
+  auto addr_arr = makeAddressArray();
+  const auto& pattern_arr = static_cast<const arrow::Array&>(
+    vast::pattern_array(vast::make_arrow_pattern(), str_arr));
+
+  CHECK(caf::get_if<arrow::StringArray>(&*str_arr));
+  CHECK(!caf::get_if<arrow::UInt64Array>(&*str_arr));
+  CHECK(!caf::get_if<arrow::StringArray>(&*uint_arr));
+  CHECK(caf::get_if<arrow::UInt64Array>(&*uint_arr));
+  CHECK(!caf::get_if<vast::address_array>(&*uint_arr));
+  CHECK(!caf::get_if<vast::pattern_array>(&*addr_arr));
+  CHECK(caf::get_if<vast::address_array>(&*addr_arr));
+  CHECK(!caf::get_if<vast::address_array>(&pattern_arr));
+  CHECK(caf::get_if<vast::pattern_array>(&pattern_arr));
+  CHECK(caf::get_if<vast::pattern_array>(&pattern_arr));
+
+  caf::visit(is_type<arrow::StringArray>(), *str_arr);
+  caf::visit(is_type<vast::pattern_array>(), pattern_arr);
+  caf::visit(is_type<vast::pattern_array>(), *str_arr);
+
+  auto f = vast::detail::overload{
+    [](const vast::address_array&) {
+      return 99;
+    },
+    [](const vast::pattern_array&) {
+      return 100;
+    },
+    [](const arrow::StringArray&) {
+      return 101;
+    },
+    [](const auto&) {
+      return -1;
+    },
+  };
+  CHECK_EQUAL(caf::visit(f, *str_arr), 101);
+  CHECK_EQUAL(caf::visit(f, pattern_arr), 100);
+  CHECK_EQUAL(caf::visit(f, *addr_arr), 99);
+  CHECK_EQUAL(caf::visit(f, *int_arr), -1);
+  caf::get_if<std::shared_ptr<vast::pattern_array>>(&str_arr);
+  CHECK(caf::visit(is_type<std::shared_ptr<arrow::StringArray>>(), str_arr));
+  CHECK(!caf::visit(is_type<std::shared_ptr<arrow::StringArray>>(), addr_arr));
+  CHECK(!caf::visit(is_type<std::shared_ptr<vast::pattern_array>>(), addr_arr));
+  CHECK(caf::visit(is_type<std::shared_ptr<vast::address_array>>(), addr_arr));
```

Comment:
Can you also add basic tests for `caf::get_if` for the `shared_ptr` versions?

Reply:
addressed